### PR TITLE
Add time window and window assigner

### PR DESCRIPTION
**Pull request by dai-chen**
Signed-off-by: Chen Dai <daichen@amazon.com>

### Description
Add window and window assigner interface along with tumbling and sliding window assigner implementation. This PR is only focus on core `WindowAssigner` interface. The future tasks in https://github.com/opensearch-project/sql/issues/954 will integrate it and other new windowing components with query plan operator.

#### Window

A Window is a time interval which has start timestamp and end timestamp. For convenience, the end timestamp is exclusive and thus the window boundary is actually `[startTime, endTime-1]`.

#### Window Assigner

A WindowAssigner identifies which window(s) the event belong. This is the first step and foundation of handling unbounded streaming data. As follows, an event may be assigned to a single window or multiple windows according to the windowing function. Here are a few example for common windowing approach.

Tumble window is fixed size without any overlap and start timestamp is aligned by mod window size.

stream-processing-Tumbling-window

Sliding window is fixed size but overlapped by slide size. The algorithm is referencing https://github.com/apache/flink/blob/master/flink-streaming-java/src/main/java/org/apache/flink/streaming/api/windowing/assigners/SlidingEventTimeWindows.java#L70

stream-processing-Sliding-window

Session windowing dynamic sized and defined by a given gap duration. It is useful for event representing user activity by analyzing user behavior within a session. Note that session window may cause window merging which will be supported in future as needed.

stream-processing-Session-window
 
### Issues Resolved
https://github.com/opensearch-project/sql/issues/951
 
### Check List
- [x] New functionality includes testing.
  - [x] All tests pass, including unit test, integration test and doctest
- [ ] New functionality has been documented.
  - [x] New functionality has javadoc added
  - [ ] New functionality has user manual doc added
- [x] Commits are signed per the DCO using --signoff 

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

File: core/src/main/java/org/opensearch/sql/planner/streaming/windowing/assigner/WindowAssigner.java [hunk 1]
```diff
@@ -0,0 +1,24 @@
...
+ * based on different windowing approach.
+ */
+public interface WindowAssigner {
+
+  /**
+   * Return window(s) assigned to the timestamp.
+   * @param timestamp given event timestamp
+   * @return windows assigned
+   */
+  List<Window> assign(long timestamp);
```
Comment (penghuo) [inline core/src/main/java/org/opensearch/sql/planner/streaming/windowing/assigner/WindowAssigner.java:22]: Set<Window>?

File: core/src/main/java/org/opensearch/sql/planner/streaming/windowing/assigner/TumblingWindowAssigner.java [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+   */
+  public TumblingWindowAssigner(long windowSize) {
+    Preconditions.checkArgument(windowSize > 0,
+        "Window size [%s] must be positive number", windowSize);
+    this.windowSize = windowSize;
+  }
+
+  @Override
+  public List<Window> assign(long timestamp) {
+    long startTime = timestamp - timestamp % windowSize;
```
Comment (penghuo) [inline core/src/main/java/org/opensearch/sql/planner/streaming/windowing/assigner/TumblingWindowAssigner.java:33]: [DateTimeUtils.](https://github.com/opensearch-project/sql/blob/2.x/core/src/main/java/org/opensearch/sql/utils/DateTimeUtils.java#L24) reuse utils function?

File: core/src/main/java/org/opensearch/sql/planner/streaming/windowing/assigner/SlidingWindowAssigner.java [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+
+    // Assign window from the last start time to first until given timestamp outside current window
+    long startTime = timestamp - timestamp % slideSize;
+    for (Window win = window(startTime); win.maxTimestamp() >= timestamp; win = window(startTime)) {
+      windows.add(win);
+      startTime -= slideSize;
+    }
+
+    // Reverse the window list for easy read and test
+    Collections.reverse(windows);
```
Comment (penghuo) [inline core/src/main/java/org/opensearch/sql/planner/streaming/windowing/assigner/SlidingWindowAssigner.java:50]: nit, does LinkedList.addFirst() help?

**Review by penghuo — APPROVED** [review]

File: core/src/main/java/org/opensearch/sql/planner/streaming/windowing/assigner/TumblingWindowAssigner.java (hunk 2 above)
Comment (dai-chen) [reply core/src/main/java/org/opensearch/sql/planner/streaming/windowing/assigner/TumblingWindowAssigner.java:33]: Good idea.

File: core/src/main/java/org/opensearch/sql/planner/streaming/windowing/assigner/TumblingWindowAssigner.java [hunk 4]
```diff
@@ -0,0 +1,36 @@
...
+ * A tumbling window assigner assigns a single window per event timestamp
+ * without overlap.
+ */
+public class TumblingWindowAssigner implements WindowAssigner {
+
+  /** Window size in millisecond. */
+  private final long windowSize;
+
+  /**
+   * Create tumbling window assigner with the given window size.
```
[the remark this answers is not in the document]
Comment (dai-chen) [reply core/src/main/java/org/opensearch/sql/planner/streaming/windowing/assigner/TumblingWindowAssigner.java:23]: In ms. Updating the Javadoc with more details.

File: core/src/main/java/org/opensearch/sql/planner/streaming/windowing/assigner/SlidingWindowAssigner.java (hunk 3 above)
Comment (dai-chen) [reply core/src/main/java/org/opensearch/sql/planner/streaming/windowing/assigner/SlidingWindowAssigner.java:50]: Sure, will change. Thanks!

**Pull request merged**